src/IP.php

```diff
@@ -272,7 +391,24 @@ public function isIn($block)
         return $block->contains($this);
     }
 
-    abstract public function isPrivate();
+    /**
+     * Return true if the address is reserved per IANA IPv4/6 Special Registry.
+     *
+     * @return bool
+     */
+    public function isPrivate()
```

Comment:
I think this method can be shorter and made faster by exiting early if a private range is found:

```
if ($this->is_private === null) {
    $this->is_private = false;
    foreach (static::$privateRanges as $range) {
        if ($this->isIn($range)) {
            $this->is_private = true;
            break;
       }
    }
}

return $this->is_private;
```

Not tested, but please have a look.

Reply:
99% of programming is good looking code ;)
I don't like your implementation here because you have an if statement wrapped in a foreach loop wrapped in an if statement.
I agree with part of your approach, that I should break out of the loop if it is in range. So I will adjust.

Reply:
Alright so you don't like if-foreach-if because it's not "good looking", but you are ok with 3 returns at 3 different places, including one return that *also* does an assignment? We don't have the same definition of good looking I guess. :-)

You can leave the first if separate if you want, I do not really mind about this one. However I do not like assignment inside return (`return $this->is_private = true;`) - to me it's the same when people do assignments inside an `if` or other shenanigans like that, it just creates unnecessary risks of confusion. Please change this line.